### PR TITLE
GEODE-10068: Make WanCopyRegionFunctionService thread pool configurab…

### PR DESCRIPTION
**Pull request by albertogpz**
…le through property

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/lang/SystemPropertyHelper.java
Comment (kohlmu-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/lang/SystemPropertyHelper.java:121]: Just to make sure that we don't get the property wrong, I would want to make sure that this property is `geode.wan.copy-region.max-threads`. So we might require a few more contextual property "levels".

@upthewaterspout @pivotal-jbarrett what are your thoughts here? I know that we've never really formally discussed contextual property hierarchies, but is this something that we might consider starting here?

File: geode-wan/src/main/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionService.java
Comment (kohlmu-pivotal) [inline geode-wan/src/main/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionService.java:50]: I seem to be missing some tests here. Or maybe there is one already confirming concurrent threads. Is this something that we want to test out? The behavior when changing the max threads setting? Is there any potential for failure if we run with 1 thread vs 10 threads vs 100 threads?

Also, I don't know how I feel if we have an unlimited thread count setting? Maybe we can initially cap the max threads? Or maybe have another (internal) property that would allow us to change the max-thread cap?

**Review by kohlmu-pivotal — CHANGES_REQUESTED** [review]

File: geode-wan/src/main/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionService.java
Comment (albertogpz) [reply geode-wan/src/main/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionService.java]: I have added a test case to make sure that when the maximum limit is reached the new command will wait for a thread to be free.
I am not sure if it is really necessary to add a cap for max threads.

Comment (kohlmu-pivotal) [reply geode-wan/src/main/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionService.java]: @albertogpz Max threads makes sure that we don't overwhelm a system. As this command can be run at any time (there is no notion of a maintenance mode for Geode) we have to assume that there would be a runtime system impact as we increase the number of threads running a region-copy.

**Comment by kohlmu-pivotal** [comment]
@albertogpz you don't need my approval... if you need to get this completed, move it out of the draft status and have the Code Owners review it.

**Comment by albertogpz** [comment]
> @albertogpz you don't need my approval... if you need to get this completed, move it out of the draft status and have the Code Owners review it.

@kohlmu-pivotal thanks for telling me. I had forgotten that it was still in draft status.

File: geode-core/src/main/java/org/apache/geode/internal/lang/SystemPropertyHelper.java
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/lang/SystemPropertyHelper.java:122]: Raised this concern in another PR about the ongoing inconsistency in naming conventions for system properties. I don't have a good answer, just raising this again.

File: geode-wan/src/main/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionService.java
Comment (pivotal-jbarrett) [inline geode-wan/src/main/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionService.java:55]: Why is this not a constant member variable?

File: geode-wan/src/main/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionService.java
Comment (pivotal-jbarrett) [inline geode-wan/src/main/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionService.java:53]: As a practice we aren't adding this annotation to package private methods used in testing.

File: geode-wan/src/test/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionServiceTest.java
Comment (pivotal-jbarrett) [inline geode-wan/src/test/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionServiceTest.java:29]: Please upgrade to JUnit 5.

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]

File: geode-core/src/main/java/org/apache/geode/internal/lang/SystemPropertyHelper.java
Comment (kohlmu-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/lang/SystemPropertyHelper.java:122]: @pivotal-jbarrett can you please elude to what the inconsistencies are and how you believe they should be resolved. (possibly even a link to a doc and agreed naming approach) As I had previously asked @albertogpz to change it to something more contextual and it seems believe it to be inconsistent. 
In addition you should see that I had previously (on 8 Mar) asked  yourself and @upthewaterspout to comment on the system properties naming convention.

Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/internal/lang/SystemPropertyHelper.java:122]: Here is the other PR #7517 currently under review. It uses the old `GeodeGlossary.GEMFIRE_PREFIX` prefix with `Class.CONSTANT` suffix. It also directly access System Properties directly via `Integer.getInteger`. This PR uses a hardcoded `geode` prefix and utilizes `SystemProperty.getProductIntegerProperty`. I haven't had time to investigate a better and consistent approach to naming and access. Just raising it again here some maybe someone will and post a proposal to dev@geode.

File: geode-wan/src/main/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionService.java
Comment (albertogpz) [reply geode-wan/src/main/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionService.java:55]: Changed

**Review by pivotal-jbarrett — APPROVED** [review]

**Review by kirklund — APPROVED** [review]

**Comment by albertogpz** [comment]
> > @albertogpz you don't need my approval... if you need to get this completed, move it out of the draft status and have the Code Owners review it.
> 
> @kohlmu-pivotal thanks for telling me. I had forgotten that it was still in draft status.

@kohlmu-pivotal It seems that I need now also your approval because you requested changes.

**Review by kohlmu-pivotal — APPROVED** [review]

**Pull request merged**
